src/Mod/Part/Gui/TaskAttacher.cpp

```diff
@@ -345,47 +350,194 @@ bool TaskAttacher::updatePreview()
+        // - - LinkToPart2
+        // - - - Cube
+        // - - - Sketch
+        obj = obj->getLinkedObject();
+
+        if (obj == group) {
+            ++i;
+            obj = doc->getObject(names[i].c_str());
+            if (!obj) {
+                return;
+            }
+
+            rootObj = obj;
+
+            // Rebuild 'sub' starting from the next element after the current 'name'
+            sub = "";
+            for (size_t j = i + 1; j < names.size(); ++j) {
+                sub += names[j];
+                if (j != names.size() - 1) {
+                    sub += ".";  // Add a period between elements
+                }
+            }
+            return;
+        }
+    }
+
+    // if we reach this point it means that attaching object's group is outside of
+    // the scope of the attached object. For instance: 
+    // - Part1
+    // - - Part2
+    // - - - Cube
+    // - - Part3
+    // - - - Sketch
+    // In this case the selection is not acceptable.
+    rootObj = nullptr;
+}
+
+void TaskAttacher::handleInitialSelection()
+{
+    // We handle initial selection only if it is not attached yet.
+    App::DocumentObject* obj = ViewProvider->getObject();
+    Part::AttachExtension* pcAttach = obj->getExtensionByType<Part::AttachExtension>();
+    std::vector<App::DocumentObject*> refs = pcAttach->AttachmentSupport.getValues();
+
+    if (!refs.empty()) {
+        return;
+    }
+    std::vector<std::string> objNames;
+    std::vector<std::string> subNames;
+
+    auto sel = Gui::Selection().getSelectionEx("",
+        App::DocumentObject::getClassTypeId(), Gui::ResolveMode::NoResolve);
+    for (auto& selObj : sel) {
+        std::vector<std::string> subs = selObj.getSubNames();
+        const char* objName = selObj.getFeatName();
+        for (auto& sub : subs) {
+            objNames.push_back(objName);
+            subNames.push_back(sub);
+        }
     }
+    addToReference(objNames, subNames);
 }
 
 void TaskAttacher::onSelectionChanged(const Gui::SelectionChanges& msg)
 {
-    if (!ViewProvider)
+    if (!ViewProvider) {
         return;
+    }
 
     if (msg.Type == Gui::SelectionChanges::AddSelection) {
-        if (iActiveRef < 0)
-            return;
+        addToReference(msg.pObjectName, msg.pSubName);
+    }
+}
 
-        // Note: The validity checking has already been done in ReferenceSelection.cpp
-        Part::AttachExtension* pcAttach = ViewProvider->getObject()->getExtensionByType<Part::AttachExtension>();
+void TaskAttacher::addToReference(std::vector<std::string> objNames, std::vector<std::string> subNames)
+{
+    if (iActiveRef < 0 || objNames.size() != subNames.size()) {
+        return;
+    }
+
+    // Note: The validity checking has already been done in ReferenceSelection.cpp
+    App::DocumentObject* obj = ViewProvider->getObject();
+    Part::AttachExtension* pcAttach = obj->getExtensionByType<Part::AttachExtension>();
+
+    for (size_t i = 0; i < objNames.size(); ++i) {

```

Comment:
Why use the C-style `for` loop here?

Reply:
To access subNames[i] 😄 
Ok perhaps we need a struct to hold obj-sub.

Reply:
I have changed to 

```
    struct ObjSubNames {
        std::string objName;
        std::string subName;
    };
    void addToReference(ObjSubNames objSubName);
    void addToReference(const std::vector<ObjSubNames>& objSubNames);
```

Reply:
A bit of nitpicking, but we should avoid plural names for structs, even if in this case it really contains multiple names. Even in the code you posted there is a problem as the type suggest that there are multiple names and variable name suggest that it is singular.  Maybe `SubReference` would be a good name?

Reply:
Ah yes indeed. I put plural because there are 2 names but it's confusing. Let's go with SubAndObjNamePair then